### PR TITLE
Make additional function capture more precise

### PR DESCRIPTION
**Pull request by cblappert**
This PR attempts to resolve issue #1140 #1093 making additional functions capture precisely the values the rewritten additional function will need. 

This is accomplished by visiting the additional function's effects when we encounter a `visitValueFunction` corresponding to that additional function. 

All values that need to be serialized in the parent scope of the additional function:
1. Object/Function captures that are
a. from the additional function or nested functions 
b. not in CreatedObjects (i.e. created in a parent scope)
2. additional function's captures that will be used by rewritten function
These are re-visited from the function's parent scope to make sure that we record the scopes for the bindings correctly.

`reVisitSet` ensures that Objects/Functions that need to be in a parent scope get placed in the correct parent scope by appending them to the common generator (where `_getTarget` would emit them). The other reason for this is that while serializing additional functions, they don't have access to previous generators/bodies. 

Other changes:
- `test-runner.js` now tests simpleClosures option. This makes test-serializer take a long time so 
- there is now a `--fast` option as well that runs each test only once.
- `test-runner.js` also has more error output giving actual stacktraces.
- added tests
- moved specialized binding functions into `visitBinding`
- add `react-mocks` compatibility option to CLI to be able to debug `test-react` with cli

### CODE REVIEW

File: src/serializer/ResidualFunctions.js [hunk 1]
```diff
@@ -525,6 +525,7 @@ export class ResidualFunctions {
               unstrictFunctionBodies.push(funcNode);
             }
           } else {
+            // TODO: make sure things here are emitted correctly
```
Comment (NTillmann) [inline src/serializer/ResidualFunctions.js]: What's wrong?

File: src/serializer/ResidualHeapSerializer.js [hunk 2]
```diff
@@ -1125,6 +1125,7 @@ export class ResidualHeapSerializer {
     for (let [boundName, residualBinding] of residualBindings) {
       let referencedValues = [];
       let serializeBindingFunc;
+      invariant(residualBinding);
```
Comment (NTillmann) [inline src/serializer/ResidualHeapSerializer.js]: Why?

File: test/serializer/basic/CircularFuncs.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// serialized function clone count: 0
```
Comment (yinghuitan) [inline test/serializer/basic/CircularFuncs.js]: Is this testing additional function? If so, I was expecting some annotations to specify target additional function.

File: src/serializer/ResidualHeapVisitor.js [hunk 4]
```diff
@@ -582,98 +589,131 @@ export class ResidualHeapVisitor {
...
+      functionValue,
+      captures: functionInfo.unbound,
+      modifiedBindings: modifiedBindingInfo,
+      instance: funcInstance,
+    });
+    this.visitGenerator(generator);
+    this._withScope(generator, visitPropertiesAndBindings);
+    // Re-visit any bindings corresponding to unbound values to make sure that
+    // they're serialized in the correct scope
+    parentScopeVisitQueue.push(() => {
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: Not sure what the exact reason for this `parentScopeVisitQueue` thing is. Instead, could you pull the `this.realm.evaluateAndRevertInGlobalEnv(` inside of this function, and then run the code you pass to `parentScopeVisitQueue` at the very end of this function?

File: src/serializer/ResidualHeapVisitor.js [hunk 5]
```diff
@@ -359,46 +343,80 @@ export class ResidualHeapVisitor {
           );
         }
       }
+      this.functionInfos.set(code, functionInfo);
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: Why did you move this?

File: src/serializer/ResidualHeapVisitor.js [hunk 6]
```diff
@@ -319,11 +299,18 @@ export class ResidualHeapVisitor {
     invariant(val instanceof ECMAScriptSourceFunctionValue);
     invariant(val.constructor === ECMAScriptSourceFunctionValue);
     let formalParameters = val.$FormalParameters;
-    invariant(formalParameters != null);
+    invariant(formalParameters !== null);
     let code = val.$ECMAScriptCode;
-    invariant(code != null);
+    invariant(code !== null);
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: Invariant shouldn't be needed anymore at all, as Flow type guarantees this.

File: src/serializer/ResidualHeapVisitor.js [hunk 7]
```diff
@@ -319,11 +299,18 @@ export class ResidualHeapVisitor {
     invariant(val instanceof ECMAScriptSourceFunctionValue);
     invariant(val.constructor === ECMAScriptSourceFunctionValue);
     let formalParameters = val.$FormalParameters;
-    invariant(formalParameters != null);
+    invariant(formalParameters !== null);
```
Comment (NTillmann) [inline src/serializer/ResidualHeapVisitor.js]: Invariant shouldn't be needed anymore at all, as Flow type guarantees this.

**Comment by NTillmann** [comment]
I high-level description of the change would be nice. Not sure what to expect...

File: src/serializer/ResidualHeapVisitor.js (hunk 4 above)
Comment (cblappert) [reply src/serializer/ResidualHeapVisitor.js]: I could, I just thought it might be a little cleaner to not have a giant anonymous function.

**Comment by cblappert** [comment]
Give me a few minutes to fix the history by rebasing instead of merging

File: src/utils/generator.js [hunk 8]
```diff
@@ -61,7 +61,7 @@ export type GeneratorBuildNodeFunction = (Array<BabelNodeExpression>, Serializat
 export type GeneratorEntry = {
   declared?: AbstractValue,
   args: Array<Value>,
-  buildNode: GeneratorBuildNodeFunction,
+  buildNode?: GeneratorBuildNodeFunction,
```
Comment (hermanventer) [inline src/utils/generator.js]: Please add a comment here to explain the significance of making the buildNode undefined.

File: src/serializer/ResidualHeapVisitor.js [hunk 9]
```diff
@@ -113,6 +115,14 @@ export class ResidualHeapVisitor {
   shouldVisitReactLibrary: boolean;
   reactElementEquivalenceSet: ReactElementSet;
 
+  // We only want to add to reVisitSet when we're in an additional function
+  inAdditionalFunction: boolean;
+  // Tracks objects + functions that were visited from inside additional functions that need to be serialized in a
+  // parent scope of the additional function (e.g. functions/objects only used from additional functions that were
+  // declared outside the additional function need to be serialized in the additional function's parent scope for
+  // identity to work).
+  reVisitSet: Set<ObjectValue>;
```
Comment (hermanventer) [inline src/serializer/ResidualHeapVisitor.js]: Maybe call this "additonalRoots"?

File: src/serializer/ResidualHeapVisitor.js [hunk 10]
```diff
@@ -695,10 +729,53 @@ export class ResidualHeapVisitor {
...
+          let residualBinding = this.visitBinding(functionValue, innerName, false);
+          if (residualBinding) {
+            residualBinding.modified = true;
+            funcInstance.residualFunctionBindings.set(innerName, residualBinding);
+          }
+        }
+        this.reVisitSet = prevReVisit;
+      }
+    );
+    this.inAdditionalFunction = false;
```
Comment (hermanventer) [inline src/serializer/ResidualHeapVisitor.js]: This subtly assumes that additional functions can never be nested inside additional functions.
